storm-client/src/jvm/org/apache/storm/scheduler/SchedulerAssignmentImpl.java

```diff
@@ -19,36 +19,82 @@
 
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
-//TODO: improve this by maintaining slot -> executors as well for more efficient operations
+import org.apache.storm.generated.WorkerResources;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 public class SchedulerAssignmentImpl implements SchedulerAssignment {
+    private static final Logger LOG = LoggerFactory.getLogger(SchedulerAssignmentImpl.class);
+
     /**
      * topology-id this assignment is for.
      */
-    String topologyId;
+    private final String topologyId;
+
     /**
      * assignment detail, a mapping from executor to <code>WorkerSlot</code>
      */
-    Map<ExecutorDetails, WorkerSlot> executorToSlot;
-    
-    public SchedulerAssignmentImpl(String topologyId, Map<ExecutorDetails, WorkerSlot> executorToSlots) {
-        this.topologyId = topologyId;
-        this.executorToSlot = new HashMap<>(0);
-        if (executorToSlots != null) {
-            this.executorToSlot.putAll(executorToSlots);
+    private final Map<ExecutorDetails, WorkerSlot> executorToSlot = new HashMap<>();
+    private final Map<WorkerSlot, WorkerResources> resources = new HashMap<>();
+    private final Map<String, Double> totalSharedOffHeap = new HashMap<>();
+
+    public SchedulerAssignmentImpl(String topologyId, Map<ExecutorDetails, WorkerSlot> executorToSlot,
+            Map<WorkerSlot, WorkerResources> resources, Map<String, Double> totalSharedOffHeap) {
+        this.topologyId = topologyId;       
+        if (executorToSlot != null) {
```

Comment:
I prefer to be very lenient with inputs if something can be null I don't want to force others to never use null.